core/src/main/java/org/apache/carbondata/core/datamap/DataMapProvider.java

```diff
@@ -264,23 +264,52 @@ private boolean getSpecificSegmentsTobeLoaded(Map<String, List<String>> segmentM
     } else {
       for (RelationIdentifier relationIdentifier : relationIdentifiers) {
         List<String> dataMapTableSegmentList = new ArrayList<>();
+        // Get all segments for parent relationIdentifier
+        List<String> mainTableSegmentList =
+            DataMapUtil.getMainTableValidSegmentList(relationIdentifier);
+        boolean ifTableStatusUpdateRequired = false;
         for (LoadMetadataDetails loadMetaDetail : listOfLoadFolderDetails) {
           if (loadMetaDetail.getSegmentStatus() == SegmentStatus.SUCCESS
               || loadMetaDetail.getSegmentStatus() == SegmentStatus.INSERT_IN_PROGRESS) {
             Map<String, List<String>> segmentMaps =
                 DataMapSegmentStatusUtil.getSegmentMap(loadMetaDetail.getExtraInfo());
+            for (String segmentId : mainTableSegmentList) {
+              // In case if dataMap segment(0) is mapped to mainTable segments{0,1,2} and if
+              // {0,1,2} segments of mainTable are compacted to 0.1. Then,
+              // on next rebuild/load to dataMap, no need to load segment(0.1) again. Update the
+              // segmentMapping of dataMap segment from {0,1,2} to {0.1}
+              if (!checkIfSegmentsToBeReloaded(relationIdentifier, segmentMaps.get(
+                  relationIdentifier.getDatabaseName() + CarbonCommonConstants.POINT
+                      + relationIdentifier.getTableName()), segmentId)) {
```

Comment:
This key is repeating in multiple places, create once and use

Reply:
done